tensorboard/plugins/graph/graph_util_test.py

```diff
@@ -0,0 +1,488 @@
+      }
+      node {
+        name: "B"
+        op: "Input"
+      }
+      node {
+        name: "C"
+        op: "MatMul"
+        input: "A"
+        input: "B"
+      }
+      versions {
+        producer: 21
+      }
+    ''', graph_def_b)
+
+    self.assertProtoEquals(
+        expected_proto,
+        graph_util.combine_graph_defs(graph_def_a, graph_def_b))
+
+  def test_combine_graph_defs_name_collided_but_same_content(self):
+    expected_proto = '''
+      node {
+        name: "X"
+        op: "Input"
+      }
+      node {
+        name: "W"
+        op: "Input"
+      }
+      node {
+        name: "Y"
+        op: "MatMul"
+        input: "X"
+        input: "W"
+      }
+      node {
+        name: "A"
+        op: "Input"
+      }
+      versions {
+        producer: 21
+      }
+    '''
+
+    graph_def_a = GraphDef()
+    text_format.Merge('''
+      node {
+        name: "X"
+        op: "Input"
+      }
+      node {
+        name: "W"
+        op: "Input"
+      }
+      node {
+        name: "Y"
+        op: "MatMul"
+        input: "X"
+        input: "W"
+      }
+      versions {
+        producer: 21
+      }
+    ''', graph_def_a)
+
+    graph_def_b = GraphDef()
+    text_format.Merge('''
+      node {
+        name: "X"
+        op: "Input"
+      }
+      node {
+        name: "A"
+        op: "Input"
+      }
+      versions {
+        producer: 21
+      }
+    ''', graph_def_b)
+
+    self.assertProtoEquals(
+        expected_proto,
+        graph_util.combine_graph_defs(graph_def_a, graph_def_b))
+
+  def test_combine_graph_defs_name_collided_differnt_content(self):
+    graph_def_a = GraphDef()
+    text_format.Merge('''
+      node {
+        name: "X"
+        op: "Input"
+      }
+      node {
+        name: "W"
+        op: "Input"
+      }
+      node {
+        name: "Y"
+        op: "MatMul"
+        input: "X"
+        input: "W"
+      }
+      versions {
+        producer: 21
+      }
+    ''', graph_def_a)
+
+    graph_def_b = GraphDef()
+    text_format.Merge('''
+      node {
+        name: "X"
+        op: "Input"
+        device: "cpu:0"
+      }
+      node {
+        name: "Z"
+        op: "Input"
+      }
+      node {
+        name: "Q"
+        op: "MatMul"
+        input: "X"
+        input: "Z"
+      }
+      versions {
+        producer: 21
+      }
+    ''', graph_def_b)
+
+    with six.assertRaisesRegex(
+        self,
+        ValueError,
+        r'Cannot combine GraphDefs because nodes share a name but are different. X'):
+      graph_util.combine_graph_defs(graph_def_a, graph_def_b)
+
+  def test_combine_graph_defs_function(self):
+    expected_proto = '''
+      library {
+        function {
+          signature {
+            name: "foo"
+            input_arg {
+              name: "x"
+              type: DT_HALF
+            }
+            output_arg {
+              name: "identity"
+              type: DT_HALF
+            }
+          }
+          node_def {
```

Comment:
Not super salient to the test, but is this a valid function definition?
The actual node def takes `x: f16` and `y: f16` to `Add: f16`, but the
signature is `(x: f16) -> identity: f16`.

Reply:
Structure-wise it is a valid but it would never be produced by TF :)
Did not focus too much on TF correctness.

Reply:
Cool; that is fine with me.